plugins/mysql/mysqlembed.cpp

```diff
@@ -93,27 +93,120 @@ static void fail(const char *message)
+            cacheOptions = strdup(_cacheOptions);
+        else
+            cacheOptions = nullptr;
     }
     ~MySQLConnection()
     {
         if (conn)
-            mysql_close(conn);
+        {
+            if (threadCached || globalCached)
+            {
+                Owned<MySQLConnection> cacheEntry = new MySQLConnection(*this);
+                if (threadCached)
+                    setThreadCache(cacheEntry.getLink());
+                else // globalCached
+                {
+                    CriticalBlock b(globalCacheCrit);
+                    if (globalCachedConnections.length()==MAX_GLOBAL_CACHE)
+                    {
+                        MySQLConnection &goer = globalCachedConnections.popGet();
+                        goer.globalCached = false;  // Make sure we don't recache it!
+                        goer.Release();
+                    }
+                    globalCachedConnections.add(*cacheEntry.getClear(), 0);
+                }
+            }
+            else
+            {
+                mysql_close(conn);
+                free((char *) cacheOptions);
+            }
+        }
     }
     inline operator MYSQL *() const
     {
         return conn;
     }
+    inline bool matches (const char *_options)
+    {
+        return _options && cacheOptions && streq(_options, cacheOptions);
+    }
+
+    static MySQLConnection *findCachedConnection(const char *options, bool threadCache, bool globalCache)
+    {
+        if (threadCache)
+        {
+            if (threadCachedConnection && threadCachedConnection->matches(options))
+            {
+                MySQLConnection *ret = threadCachedConnection;
+                threadCachedConnection = nullptr;
+                return ret;
+            }
+        }
+        else if (globalCache)
+        {
+            CriticalBlock b(globalCacheCrit);
+            ForEachItemIn(idx, globalCachedConnections)
+            {
+                MySQLConnection &cached = globalCachedConnections.item(idx);
+                if (cached.matches(options))
+                {
+                    globalCachedConnections.remove(idx, true);
+                    return &cached;
+                }
+            }
+        }
+        return nullptr;
+    }
+
+    static void clearThreadCache()
+    {
+        ::Release(threadCachedConnection);
+        threadCachedConnection = NULL;
+    }
+
+    static void setThreadCache(MySQLConnection *connection)
+    {
+        clearThreadCache();
+        threadCachedConnection = connection;
+    }
+
 private:
-    MySQLConnection(const MySQLConnection &);
+    MySQLConnection(const MySQLConnection &from)
+    {
+        conn = from.conn;
+        cacheOptions = from.cacheOptions;
+        threadCached = from.threadCached;
+        globalCached = from.globalCached;
+    }
+
+    static CIArrayOf<MySQLConnection> globalCachedConnections;
+    static CriticalSection globalCacheCrit;
+
     MYSQL *conn;
+    const char *cacheOptions;
```

Comment:
It is strdup'ed in the constructor. The intention is to avoid re-duping in the copy constructor (which is used when taking over ownership from a dying object). The code is not especially clear though - I'm open to suggestions about how to code it more cleanly.